### ISSUE TITLE
bug: fibonacci evaluation errors on first step on `dst_addr` lookup

### ISSUE THREAD

**Issue by jobez** [keep-starknet-strange/cairo-zig#205]
See https://github.com/keep-starknet-strange/cairo-zig/actions/runs/7055855307/job/19206936981?pr=157 for the failing case. Locally I compared our evaluation (right side) to the cairo-vm_in_go (left side) at commit e806055 (closer to the fibonacci milestone) and was seeing this difference:

![telegram-cloud-photo-size-1-5132342471563520857-y](https://github.com/keep-starknet-strange/cairo-zig/assets/615197/3aa25251-b32f-4ee9-85c2-1876a6cf3432)

**Comment by jobez** [comment]
Some debugging notes to make sure I am understanding why logic is there so I can ensure any changes in getting one instruction to be able to be ran in fib does not produce regressions elsehwere:

# Looking specifically into `error.MemoryOutOfBounds`

The lines we are investigating are these, essentially, as this is where the error in question is being thrown/return:

https://github.com/keep-starknet-strange/cairo-zig/blob/3b937719cd0e4e3dee8e17bc4f9088d322dec622/src/vm/memory/memory.zig#L408-L410

I assume the `memory.get` logic is following the rust implementation https://github.com/lambdaclass/cairo-vm/blob/main/vm/src/vm/vm_memory/memory.rs#L180, 

in comparison to the cairovm rust impl, that calls into 'get' that will return nil if out of bounds https://doc.rust-lang.org/src/core/slice/mod.rs.html#609

but on inspection I am not following why we have specific invalidSegment/Index checks, other than maybe different semantics for zig array list.

tracking the commits:
https://github.com/keep-starknet-strange/cairo-zig/commit/db169a9a3e148c973efdc06d16a8f3b1a02bf935

the index checks were atleast a refactor of the previous commit but originated here:
https://github.com/keep-starknet-strange/cairo-zig/commit/a6689d6d0bf052fb8c8db6f61c169491c5ebaa68#diff-549793bb1da89b52966c6ca1f5f6ebe596c6b261a6f8f9df1d08ce39a7d10ba7R255

tl;dr Just wanted to give space if anyone understood the motivation for these lines and if not, let us lay them to rest:
https://github.com/keep-starknet-strange/cairo-zig/blob/3b937719cd0e4e3dee8e17bc4f9088d322dec622/src/vm/memory/memory.zig#L408-L410
